src/lib/mdns/MdnsCache.h

```diff
@@ -0,0 +1,183 @@
+        elementsUsed++;
+
+        return CHIP_NO_ERROR;
+    }
+
+    CHIP_ERROR Delete(PeerId peerId)
+    {
+        MdnsCacheEntry * pentry;
+        const uint64_t currentTime = mTimeSource.GetCurrentMonotonicTimeMs();
+
+        VerifyOrReturnError(pentry = FindPeerId(peerId, currentTime), CHIP_ERROR_KEY_NOT_FOUND);
+
+        MarkEntryUnused(*pentry);
+        return CHIP_NO_ERROR;
+    }
+
+    // given a peerId, find the parameters if its in the cache, or return error
+    CHIP_ERROR Lookup(PeerId peerId, Inet::IPAddress & addr, uint16_t & port, Inet::InterfaceId & iface)
+    {
+        MdnsCacheEntry * pentry;
+        const uint64_t currentTime = mTimeSource.GetCurrentMonotonicTimeMs();
+
+        VerifyOrReturnError(pentry = FindPeerId(peerId, currentTime), CHIP_ERROR_KEY_NOT_FOUND);
+
+        addr  = pentry->ipAddr;
+        port  = pentry->port;
+        iface = pentry->ifaceId;
+
+        return CHIP_NO_ERROR;
+    }
+
+    // only useful if MDNS_LOGGING is set.   If not used, should be optimized out
+    void DumpCache()
+    {
+        int i = 0;
+
+        MdnsLogProgress(Discovery, "cache size = %d", elementsUsed);
+        for (MdnsCacheEntry & e : mLookupTable)
+        {
+            if (e.peerId == nullPeerId)
+            {
+                MdnsLogProgress(Discovery, "Entry %d unused", i);
+            }
+            else
+            {
+                char address[100];
+
+                e.ipAddr.ToString(address, sizeof address);
+                MdnsLogProgress(Discovery, "Entry %d: node %lx fabric %lx, port = %d, address = %s", i, e.peerId.GetNodeId(),
+                                e.peerId.GetFabricId(), e.port, address);
+            }
+            i++;
+        }
+    }
+
+private:
+    struct MdnsCacheEntry
+    {
+        PeerId peerId;
+        Inet::IPAddress ipAddr;
+        uint16_t port;
+        Inet::InterfaceId ifaceId;
+        uint64_t TTL;        // from mdns record -- units?
+        uint64_t expiryTime; // units?
+    };
+    PeerId nullPeerId; // indicates a cache entr is unused
+    int elementsUsed;  // running count of how many entries are used -- for a sanity check
+
+    MdnsCacheEntry mLookupTable[CACHE_SIZE];
+    Time::TimeSource<Time::Source::kSystem> mTimeSource;
+
+    MdnsCacheEntry * findSlot(uint64_t currentTime)
+    {
+        for (MdnsCacheEntry & entry : mLookupTable)
+        {
+            if (entry.peerId == nullPeerId)
+                return &entry;
+
+            if (entry.expiryTime <= currentTime)
+            {
+                MarkEntryUnused(entry);
+                return &entry;
+            }
+        }
+        return nullptr;
+    }
+
+    MdnsCacheEntry * FindPeerId(PeerId peerId, uint64_t current_time)
+    {
+        for (MdnsCacheEntry & entry : mLookupTable)
+        {
+            if (entry.peerId == peerId)
+                return &entry;
+            if (entry.peerId != nullPeerId && entry.expiryTime <= current_time)
```

Comment:
Should this condition be checked before the `entry.peerId == peerId`? Otherwise, an expired entry can be returned, right?

Reply:
Makes sense.

Here's the new code:
```

    MdnsCacheEntry * FindPeerId(PeerId peerId, uint64_t current_time)
    {
        for (MdnsCacheEntry & entry : mLookupTable)
        {
            if (entry.peerId == peerId) {
		if(entry.expiryTime < current_time) {
	            MarkEntryUnused(entry);	
                    break;	// return nullptr
		} else return &entry;
	    }
            if (entry.peerId != nullPeerId && entry.expiryTime < current_time)
            { 
                MarkEntryUnused(entry);
            }
        }

        return nullptr;
    }
```